### PR TITLE
Remove `BaseDomain` use in `LoaderAllocator` and event tracing helpers

### PR DESCRIPTION
**Pull request by elinor-fung**
- Remove `BaseDomain` member on `LoaderAllocator`
  - Add asserts in functions using `AppDomain` that the loader allocator is collectible and the type is `LAT_Assembly` (so `AssemblyLoaderAllocator` which always had `AppDomain`)
- Remove unnecessary `BaseDomain`/`AppDomain` parameters from event tracing helpers
  - They were always being called with the current app domain

cc @jkotas @AaronRobinsonMSFT

### CODE REVIEW

File: src/coreclr/inc/eventtracebase.h [hunk 1]
```diff
@@ -750,12 +749,11 @@ namespace ETW
 #ifdef FEATURE_EVENT_TRACE
         static VOID SendThreadRundownEvent();
         static VOID SendGCRundownEvent();
-        static VOID IterateDomain(BaseDomain *pDomain, DWORD enumerationOptions);
         static VOID IterateAppDomain(AppDomain * pAppDomain, DWORD enumerationOptions);
         static VOID IterateCollectibleLoaderAllocator(AssemblyLoaderAllocator *pLoaderAllocator, DWORD enumerationOptions);
         static VOID IterateAssembly(Assembly *pAssembly, DWORD enumerationOptions);
         static VOID IterateModule(Module *pModule, DWORD enumerationOptions);
-        static VOID EnumerationHelper(Module *moduleFilter, BaseDomain *domainFilter, DWORD enumerationOptions);
+        static VOID EnumerationHelper(Module *moduleFilter, AppDomain *domainFilter, DWORD enumerationOptions);
```
Comment (jkotas) [inline src/coreclr/inc/eventtracebase.h:756]: Should the `domainFilter` argument be deleted instead?

File: src/coreclr/inc/eventtracebase.h [hunk 2]
```diff
@@ -892,7 +890,7 @@ namespace ETW
         static VOID ModuleLoad(Module *pModule, LONG liReportedSharedModule);
 #else
     public:
-        static VOID DomainLoad(BaseDomain *pDomain, _In_opt_ LPWSTR wszFriendlyName=NULL) {};
+        static VOID DomainLoad(AppDomain *pDomain, _In_opt_ LPWSTR wszFriendlyName=NULL) {};
```
Comment (AaronRobinsonMSFT) [inline src/coreclr/inc/eventtracebase.h:893]: Do any of these load operations needs the `AppDomain`?

File: src/coreclr/inc/eventtracebase.h [hunk 3]
```diff
@@ -904,7 +902,7 @@ namespace ETW
     {
         friend class ETW::EnumerationLog;
 #ifdef FEATURE_EVENT_TRACE
-        static VOID SendEventsForJitMethods(BaseDomain *pDomainFilter, LoaderAllocator *pLoaderAllocatorFilter, DWORD dwEventOptions);
+        static VOID SendEventsForJitMethods(AppDomain *pDomainFilter, LoaderAllocator *pLoaderAllocatorFilter, DWORD dwEventOptions);
```
Comment (AaronRobinsonMSFT) [inline src/coreclr/inc/eventtracebase.h:905]: Remove filter too?

File: src/coreclr/vm/eventtrace.cpp [hunk 4]
```diff
@@ -3926,20 +3926,20 @@ VOID ETW::EnumerationLog::ProcessShutdown()
...
+VOID ETW::LoaderLog::SendDomainEvent(AppDomain *pDomain, DWORD dwEventOptions, LPCWSTR wszFriendlyName)
 {
     CONTRACTL {
         THROWS;
         GC_TRIGGERS;
     } CONTRACTL_END;
 
-    if(!pBaseDomain)
+    if(!pDomain)
         return;
```
Comment (AaronRobinsonMSFT) [inline src/coreclr/vm/eventtrace.cpp:3937]: Will this ever result to `true`?

File: src/coreclr/vm/loaderallocator.cpp [hunk 5]
```diff
@@ -1725,15 +1724,15 @@ void DomainAssemblyIterator::operator++()
...
+void AssemblyLoaderAllocator::Init()
 {
     m_Id.Init();
 
     // This is CRST_UNSAFE_ANYMODE to enable registering/unregistering dependent handles to native objects without changing the
     // GC mode, in case the caller requires that
     m_dependentHandleToNativeObjectSetCrst.Init(CrstLeafLock, CRST_UNSAFE_ANYMODE);
 
-    LoaderAllocator::Init((BaseDomain *)pAppDomain);
+    LoaderAllocator::Init(NULL);
```
Comment (AaronRobinsonMSFT) [inline src/coreclr/vm/loaderallocator.cpp:1735]:
```suggestion
    LoaderAllocator::Init(NULL /* pExecutableHeapMemory */);
```

**Review by jkotas — APPROVED** [review]
Thanks

**Pull request merged**